cargo-shuttle/src/factory.rs

```diff
@@ -1,14 +1,231 @@
+        {
+            trace!("DB container '{container_name}' not running, so starting it");
+            self.docker
+                .start_container(&container_name, None::<StartContainerOptions<String>>)
+                .await
+                .expect("failed to start not running container");
+        }
+
+        self.wait_for_ready(&container_name).await?;
+
+        let db_info = DatabaseReadyInfo {
+            database_name: "postgres".to_string(),
+            role_name: "postgres".to_string(),
+            role_password: PG_PASSWORD.to_string(),
+        };
+
+        let conn_str = db_info.connection_string("localhost");
+
+        println!(
+            "{:>12} can be reached at {}\n",
+            "DB ready".bold().cyan(),
+            conn_str
+        );
+
+        Ok(conn_str)
+    }
+}
+
+impl LocalFactory {
+    async fn wait_for_ready(&self, container_name: &str) -> Result<(), shuttle_service::Error> {
+        loop {
+            trace!("waiting for '{container_name}' to be ready for connections");
+
+            let config = CreateExecOptions {
+                cmd: Some(vec!["pg_isready"]),
+                attach_stdout: Some(true),
+                ..Default::default()
+            };
+
+            let CreateExecResults { id } = self
+                .docker
+                .create_exec(container_name, config)
+                .await
+                .expect("failed to create exec to check if container is ready");
+
+            let ready_result = self
+                .docker
+                .start_exec(&id, None)
+                .await
+                .expect("failed to execute ready command");
+
+            if let bollard::exec::StartExecResults::Attached { mut output, .. } = ready_result {
+                while let Some(line) = output.next().await {
+                    if let bollard::container::LogOutput::StdOut { message } =
+                        line.expect("output to have a log line")
+                    {
+                        if message.ends_with(b"accepting connections\n") {
+                            return Ok(());
+                        }
+                    }
+                }
+            }
+
+            sleep(Duration::from_millis(500)).await;
+        }
+    }
+
+    async fn pull_image(&self, image: &str) -> Result<(), String> {
+        trace!("pulling latest image for '{image}'");
+        let mut layers = Vec::new();
+
+        let create_image_options = Some(CreateImageOptions {
+            from_image: image,
+            ..Default::default()
+        });
+        let mut output = self.docker.create_image(create_image_options, None, None);
+
+        while let Some(line) = output.next().await {
+            let info = line.expect("failed to create image");
+
+            if let Some(id) = info.id.as_ref() {
+                match layers.iter_mut().find(|item: &&mut CreateImageInfo| {
+                    item.id.is_some() && item.id.as_deref().unwrap() == id
```

Comment:
I think you can do a direct comparison instead of unwrapping here.

```suggestion
                    item.id.as_deref() == Some(id)
```